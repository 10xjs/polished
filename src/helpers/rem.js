// @flow

import endsWith from '../internalHelpers/_endsWith'
import stripUnit from './stripUnit'

/**
 * Convert pixel value to rems. The default base value is 16px, but can be changed by passing a
 * second argument to the function.
 * @example
 * // Styles as object usage
 * const styles = {
 *   'height': rem('16px')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   height: ${rem('16px')}
 * `
 *
 * // CSS in JS Output
 *
 * element {
 *   'height': '1rem'
 * }
 */

/* eslint-disable no-param-reassign */
function rem(pxval: string|number, base?: string|number = '16px') {
  let newPxval = pxval
  let newBase = base
  if (typeof pxval === 'string') {
    if (!endsWith(pxval, 'px')) {
      throw new Error(`Expected a string ending in "px" or a number passed as the first argument to rem(), got "${pxval}" instead.`)
    }
    newPxval = stripUnit(pxval)
  }

  if (typeof base === 'string') {
    if (!endsWith(base, 'px')) {
      throw new Error(`Expected a string ending in "px" or a number passed as the second argument to rem(), got "${base}" instead.`)
    }
    newBase = stripUnit(base)
  }

  if (typeof newPxval === 'string') {
    throw new Error(`Passed invalid pixel value ("${pxval}") to rem(), please pass a value like "12px" or 12.`)
  }

  if (typeof newBase === 'string') {
    throw new Error(`Passed invalid base value ("${base}") to rem(), please pass a value like "12px" or 12.`)
  }

  return `${newPxval / newBase}rem`
}

export default rem
