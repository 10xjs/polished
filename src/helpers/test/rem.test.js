// @flow
import rem from '../rem'

describe('rem', function() {
  it('should convert a simple number to rems', () => {
    expect({ height: rem(16) }).toMatchSnapshot()
  })

  it('should convert a simple string with px to rems', () => {
    expect({ height: rem('16px') }).toMatchSnapshot()
  })

  it('should convert a complex number to rems', () => {
    expect({ height: rem(18) }).toMatchSnapshot()
  })

  it('should convert a complex string with px to rems', () => {
    expect({ height: rem('18px') }).toMatchSnapshot()
  })

  it('should allow changing the base with a number', () => {
    expect({ height: rem('16px', 8) }).toMatchSnapshot()
  })

  it('should allow changing the base with a string', () => {
    expect({ height: rem('16px', '8px') }).toMatchSnapshot()
  })
})
