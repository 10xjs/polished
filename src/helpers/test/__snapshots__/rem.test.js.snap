exports[`rem should allow changing the base with a number 1`] = `
Object {
  "height": "2rem",
}
`;

exports[`rem should allow changing the base with a string 1`] = `
Object {
  "height": "2rem",
}
`;

exports[`rem should convert a complex number to rems 1`] = `
Object {
  "height": "1.125rem",
}
`;

exports[`rem should convert a complex string with px to rems 1`] = `
Object {
  "height": "1.125rem",
}
`;

exports[`rem should convert a simple number to rems 1`] = `
Object {
  "height": "1rem",
}
`;

exports[`rem should convert a simple string with px to rems 1`] = `
Object {
  "height": "1rem",
}
`;
